lib/resty/core/socket.lua

```diff
@@ -0,0 +1,85 @@
+local base = require "resty.core.base"
+local debug = require 'debug'
+local ffi = require 'ffi'
+
+
+local error = error
+local tonumber = tonumber
+local registry = debug.getregistry()
+local ffi_new = ffi.new
+local ffi_string = ffi.string
+local C = ffi.C
+
+
+ffi.cdef[[
+typedef struct ngx_http_lua_socket_tcp_upstream_s
+    ngx_http_lua_socket_tcp_upstream_t;
+
+int 
+ngx_http_lua_ffi_socket_tcp_getoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    const char *opt, int *val, char **err_msg);
+
+int 
+ngx_http_lua_ffi_socket_tcp_setoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    const char *opt, int val, char **err_msg); 
+]]
+
+local output_value_buf = ffi_new("int[1]")
```

Comment:
`int` may not be safe here?

Reply:
keepalive, reuseaddr and tcp-nodelay use boolean;
rcvbuf and sndbuf are limit by  /proc/sys/net/core/rmem_max and /proc/sys/net/core/wmem_max in linux, and maximum is 2^31-1.

Reply:
better add a comment before this line, like:
```
int is safe here since the maximum is 2^31-1
```

Reply:
accept

Reply:
accept

Reply:
The man page actually states it should be an int value for those setsockopt options we just implemented. So the value range is not the key, but the value type is.